libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/RoutingTask.kt

```diff
@@ -2,49 +2,84 @@ package net.cyclestreets.routing
 
 import android.content.Context
 import android.os.AsyncTask
+import android.util.Log
 import android.widget.Toast
+import net.cyclestreets.CycleStreetsPreferences
+import net.cyclestreets.RoutePlans.PLAN_LEISURE
 
 import net.cyclestreets.api.JourneyPlanner
 import net.cyclestreets.content.RouteData
 import net.cyclestreets.util.Dialog
+import net.cyclestreets.util.Logging
 import net.cyclestreets.util.ProgressDialog
+import net.cyclestreets.view.R
 
 abstract class RoutingTask<Params> protected constructor(private val initialMsg: String,
                                                          private val context: Context) : AsyncTask<Params, Int, RouteData>() {
     private var progress: ProgressDialog? = null
     private var error: String? = null
+    private val TAG = Logging.getTag(RoutingTask::class.java)
 
     protected constructor(progressMessageId: Int,
                           context: Context) : this(context.getString(progressMessageId), context)
 
-    protected fun fetchRoute(routeType: String,
-                             speed: Int,
-                             waypoints: Waypoints): RouteData? {
-        return fetchRoute(routeType, -1, speed, waypoints)
-    }
-
-    @JvmOverloads
     protected fun fetchRoute(routeType: String,
                              itinerary: Long,
                              speed: Int,
-                             waypoints: Waypoints? = null): RouteData? {
+                             waypoints: Waypoints? = null,
+                             distance: Int? = null,
+                             duration: Int? = null,
+                             pois: String? = null): RouteData? {
         return try {
-            val json = doFetchRoute(routeType, itinerary, speed, waypoints)
-            RouteData(json, waypoints, null)
+            val json = doFetchRoute(routeType, itinerary, speed, waypoints, distance, duration, pois)
+
+            when {
+                (json == null) -> {
+                    error = context.getString(R.string.route_not_found)
+                    throw Exception(error)
```

Comment:
Don't throw raw `Exception`, please 🤮 
If you deliberately want to use a checked exception, then it must be a specific custom subclass, so we can explicitly handle it in a specific way.

As you're just using it to jump to the catch block, I'd accept throwing raw `RuntimeException`, if you don't want to define a custom exception.

Reply:
I must admit, I only covered error handling and exceptions briefly in my reading so far, so this is the perfect cue for me to go and do some more in-depth reading.  Bear with me... :-)

Reply:
I've now defined a custom exception but not sure if this was what you had in mind?